src/storage/v2/constraints/type_constraints.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright 2024 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "type_constraints.hpp"
+#include <optional>
+#include "storage/v2/constraints/type_constraints_type.hpp"
+#include "storage/v2/property_value.hpp"
+#include "utils/algorithm.hpp"
+#include "utils/logging.hpp"
+
+namespace memgraph::storage {
+
+[[nodiscard]] std::optional<ConstraintViolation> TypeConstraints::ValidateVertexOnConstraint(const Vertex &vertex,
+                                                                                             LabelId label,
+                                                                                             PropertyId property,
+                                                                                             TypeConstraintsType type) {
+  if (vertex.deleted || !utils::Contains(vertex.labels, label)) {
+    return std::nullopt;
+  }
+  auto prop_value = vertex.properties.GetProperty(property);
```

Comment:
Can use GetPropertyOfTypes to avoid hydrating properties of types we don't care about.